crates/client-api-messages/src/websocket.rs

```diff
@@ -0,0 +1,326 @@
+#[sats(crate = spacetimedb_lib)]
+pub struct ReducerCallInfo {
+    /// The name of the reducer that was called.
+    pub reducer_name: String,
+    /// The numerical id of the reducer that was called.
+    pub reducer_id: u32,
+    /// The arguments to the reducer, encoded as BSATN or JSON according to the reducer's argument schema
+    /// and the client's requested protocol.
+    pub args: EncodedValue,
+    /// An identifier for a client request
+    pub request_id: u32,
+}
+
+/// The status of a [`TransactionUpdate`].
+#[derive(SpacetimeType, Debug)]
+#[sats(crate = spacetimedb_lib)]
+pub enum UpdateStatus {
+    /// The reducer ran successfully and its changes were committed to the database.
+    /// The rows altered in the database/ will be recorded in the `DatabaseUpdate`.
+    Committed(DatabaseUpdate),
+    /// The reducer errored, and any changes it attempted to were rolled back.
+    /// This is the error message.
+    Failed(String),
+    /// The reducer was interrupted due to insufficient energy/funds,
+    /// and any changes it attempted to make were rolled back.
+    OutOfEnergy,
+}
+
+/// A collection of inserted and deleted rows, contained in a [`TransactionUpdate`] or [`SubscriptionUpdate`].
+#[derive(SpacetimeType, Debug, Clone, Default)]
+#[sats(crate = spacetimedb_lib)]
+pub struct DatabaseUpdate {
+    pub tables: Vec<TableUpdate>,
+}
+
+impl DatabaseUpdate {
+    pub fn is_empty(&self) -> bool {
+        self.tables.is_empty()
+    }
+
+    pub fn num_rows(&self) -> usize {
+        self.tables.iter().map(|t| t.num_rows()).sum()
+    }
+}
+
+impl FromIterator<TableUpdate> for DatabaseUpdate {
+    fn from_iter<T: IntoIterator<Item = TableUpdate>>(iter: T) -> Self {
+        DatabaseUpdate {
+            tables: iter.into_iter().collect(),
+        }
+    }
+}
+
+/// Part of a [`DatabaseUpdate`] received by client from database for alterations to a single table.
+#[derive(SpacetimeType, Debug, Clone)]
+#[sats(crate = spacetimedb_lib)]
+pub struct TableUpdate {
+    /// The id of the table. Clients should prefer `table_name`, as it is a stable part of a module's API,
+    /// whereas `table_id` may change between runs.
+    pub table_id: TableId,
+    /// The name of the table.
+    pub table_name: String,
+    /// When in a [`TransactionUpdate`], the matching rows of this table deleted by the transaction.
+    ///
+    /// Rows are encoded as BSATN or JSON according to the table's schema
+    /// and the client's requested protocol.
+    ///
+    /// Always empty when in an [`InitialSubscription`].
+    pub deletes: Vec<EncodedValue>,
+    /// When in a [`TransactionUpdate`], the matching rows of this table inserted by the transaction.
+    /// When in an [`InitialSubscription`], the matching rows of this table in the entire committed state.
+    ///
+    /// Rows are encoded as BSATN or JSON according to the table's schema
+    /// and the client's requested protocol.
+    pub inserts: Vec<EncodedValue>,
```

Comment:
I'm inclined to agree with either of the two proposed solutions above. What is currently in this PR is a bit of a performance and data regression from what we currently do in protobuf because it stores an enum tag for each row of a table. This could be significant in the case of small rows like `LocationState`.

Side note, I don't see why `num_row` is required in the first version, @Centril.

I am okay to not block the review on this, but only if there is a corresponding ticket to fix this before we stabilize the API.

Reply:
`num_rows` is necessary given the way Joshua has implemented subscription row-count metrics.

Ideally, I would like to make codegen properly support generic types, and then make `ServerMessage` and `ClientMessage` generic over a row-encoding. Failing that, I think separating `ServerMessageJson` from `ServerMessageBsatn` would be reasonable. Hoisting the tag higher up is probably the least good option, but would be fine. I appreciate making this a follow-up; I think that making changes to the WS schema will be much easier once this PR merges.